### PR TITLE
feat: find best embedding matches

### PR DESCRIPTION
**Pull request by thewh1teagle**
Resolve https://github.com/k2-fsa/sherpa-onnx/issues/1096

### CODE REVIEW

File: sherpa-onnx/c-api/c-api.h [hunk 1]
```diff
@@ -1109,6 +1109,39 @@ SHERPA_ONNX_API const char *SherpaOnnxSpeakerEmbeddingManagerSearch(
 SHERPA_ONNX_API void SherpaOnnxSpeakerEmbeddingManagerFreeSearch(
     const char *name);
 
+SHERPA_ONNX_API typedef struct SherpaOnnxSpeakerEmbeddingManagerSpeakerMatch {
+  float score;
+  char *name;
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.h]:
```suggestion
  const char *name;
```

Add `const` to indicate that users should not modify it.

File: sherpa-onnx/c-api/c-api.h [hunk 2]
```diff
@@ -1109,6 +1109,39 @@ SHERPA_ONNX_API const char *SherpaOnnxSpeakerEmbeddingManagerSearch(
...
+} SherpaOnnxSpeakerEmbeddingManagerSpeakerMatch;
+
+SHERPA_ONNX_API typedef struct
+    SherpaOnnxSpeakerEmbeddingManagerBestMatchesResult {
+  SherpaOnnxSpeakerEmbeddingManagerSpeakerMatch *matches;
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.h]:
```suggestion
  const SherpaOnnxSpeakerEmbeddingManagerSpeakerMatch *matches;
```

File: sherpa-onnx/csrc/speaker-embedding-manager.cc [hunk 3]
```diff
@@ -131,6 +131,39 @@ class SpeakerEmbeddingManager::Impl {
...
+    for (int i = 0; i < scores.size(); ++i) {
+      if (scores[i] >= threshold) {
+        score_indices.emplace_back(scores[i], i);
+      }
+    }
+
+    std::sort(score_indices.rbegin(), score_indices.rend(),
+              [](const auto &a, const auto &b) { return a.first < b.first; });
+
+    for (int i = 0; i < std::min(n, static_cast<int32_t>(score_indices.size()));
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/speaker-embedding-manager.cc:159]:
```suggestion
    matches.reserve(score_indices.size());
    for (int i = 0; i < std::min(n, static_cast<int32_t>(score_indices.size()));
```

It is good to pre-allocate space.

**Review by csukuangfj — COMMENTED** [review]
Thanks! Left some minor comments. Otherwise, it looks good to me.

File: sherpa-onnx/c-api/c-api.h (hunk 2 above)
Comment (thewh1teagle) [reply sherpa-onnx/c-api/c-api.h]: When using const there I get the following compile error

```console
/Users/user/Documents/sherpa-rs/sys/sherpa-onnx/sherpa-onnx/c-api/c-api.cc:1275:30: error: read-only variable is not assignable
    result->matches[i].score = matches[i].score;
    ~~~~~~~~~~~~~~~~~~~~~~~~ ^
/Users/user/Documents/sherpa-rs/sys/sherpa-onnx/sherpa-onnx/c-api/c-api.cc:1281:29: error: read-only variable is not assignable
    result->matches[i].name = name;
    ~~~~~~~~~~~~~~~~~~~~~~~ ^
```

I think we'll need to use internal non const in the function and case it to const before return

File: sherpa-onnx/csrc/speaker-embedding-manager.cc (hunk 3 above)
Comment (thewh1teagle) [reply sherpa-onnx/csrc/speaker-embedding-manager.cc:159]: looks better! Thanks :)

File: sherpa-onnx/c-api/c-api.h (hunk 2 above)
Comment (thewh1teagle) [reply sherpa-onnx/c-api/c-api.h]: How about


```cpp
const SherpaOnnxSpeakerEmbeddingManagerBestMatchesResult *
SherpaOnnxSpeakerEmbeddingManagerGetBestMatches(
    const SherpaOnnxSpeakerEmbeddingManager *p, const float *v, float threshold,
    int32_t n) {
  auto matches = p->impl->GetBestMatches(v, threshold, n);

  if (matches.empty()) {
    return nullptr;
  }

  auto *result =
      static_cast<SherpaOnnxSpeakerEmbeddingManagerBestMatchesResult *>(
          operator new(
              sizeof(SherpaOnnxSpeakerEmbeddingManagerBestMatchesResult)));
  auto *matchesArray =
      new SherpaOnnxSpeakerEmbeddingManagerSpeakerMatch[matches.size()];

  for (int i = 0; i < matches.size(); ++i) {
    matchesArray[i].score = matches[i].score;

    char *name = new char[matches[i].name.size() + 1];
    std::copy(matches[i].name.begin(), matches[i].name.end(), name);
    name[matches[i].name.size()] = '\0';

    matchesArray[i].name = name;
  }
  new (result) SherpaOnnxSpeakerEmbeddingManagerBestMatchesResult{
      matchesArray, static_cast<int32_t>(matches.size())};

  return result;
}
```

Comment (thewh1teagle) [reply sherpa-onnx/c-api/c-api.h]: Fixed in https://github.com/k2-fsa/sherpa-onnx/pull/1102/commits/7d2341d1fe1ca45a2890a554fe2dc5360d6f15da

**Comment by csukuangfj** [comment]
Thank you for your first-time contribution!

**Pull request merged**